# Benchmarking of methods to analyze data derived from GBS-MeDIP

This repository contains scripts used for benchmarking various statistical and bioinformatic methods specifically designed for analyzing data obtained through Genotyping-by-Sequencing combined with Methylated DNA Immunoprecipitation (GBS-MeDIP).

## Overview

GBS-MeDIP is a cost-effective method developed to investigate DNA methylation. Due to its unique data characteristics, traditional methods for RNA-seq or MeDIP-seq analyses are not suitable. This repository provides scripts used to create a count matrix and to perform differential methylated analysis specifically for GBS-MeDIP data.

## Content Description

#### 1. Count Matrix Generation

create_saf_files.py: Script for merging genomic coordinates and generating SAF files used in subsequent counting steps.

feature_counts.sh: Shell script utilizing featureCounts software to generate accurate count matrices from BAM files.

compare_counts.R: R script to visualize and compare count matrices generated by MEDIPS and featureCounts.

#### 2. Distribution Analysis

distribution_fitting.R: R script to evaluate the statistical distribution (normal, Poisson, negative-binomial) fitting the GBS-MeDIP count data. Calculates and visualizes Akaike Information Criterion (AIC) values.

#### 3. Statistical Method Benchmarking

pvalue_uniformity_test.R: Evaluates the uniformity of p-value distributions to detect biases in various statistical methods.

false_positive_rate.R: Computes the False Positive Rate (FPR) for statistical methods such as Mann-Whitney, EdgeR (Maximum and Quasi-likelihood), and Limma.

simulation_true_positive_rate.R: Simulates data to compute True Positive Rates (TPR) and generates Receiver Operating Characteristic (ROC) curves to compare the accuracy of different methods.

## Recommended Pipeline

Based on our benchmarking results, we recommend the following pipeline for GBS-MeDIP data:

Count Matrix Generation: featureCounts

Differential Methylation Analysis: Mann-Whitney test

## Dependencies

R packages: ggplot2, fitdistrplus, EdgeR, Limma, MEDIPS

Tools: featureCounts, samtools, Bowtie2

## Usage

Clone the repository and follow the script-specific instructions provided as comments within each script.

git clone https://github.com/Violeta-de-Anca/Benchmarking-of-methods-to-analyse-data-derived-from-GBS-MeDIP.git

## References

Detailed methodology, results, and discussion are presented in our published paper:

Violeta de Anca Prado, et al. "Benchmarking of methods to analyze data derived from GBS-MeDIP." (Details and DOI/link to be added upon publication).

## Authors

Violeta de Anca Prado

Fábio Pértille

Pedro Sá

Marta Gòdia

Joëlle Rüegg

Carlos Guerrero Bosagna (Corresponding author)

## Funding

This project was funded by GEroNIMO (Grant Agreement No 101000236) and Templeton Foundation (Grant ID 62167).

## Acknowledgements

We acknowledge support from the National Academic Infrastructure for Supercomputing in Sweden (NAISS), the John Templeton Foundation, the European Union, Svenska Forskningsrådet, and the GEroNIMO project workshop attendees.
